pkg/controller/cluster/cluster_controller_test.go

```diff
@@ -0,0 +1,980 @@
+func newExpectedComputeNodeGroupDeleteAction(cluster *clusteroperator.Cluster, computeName string) expectedNodeGroupDeleteAction {
+	return expectedNodeGroupDeleteAction{
+		namePrefix: getNamePrefixForComputeNodeGroup(cluster, computeName),
+	}
+}
+
+// expectedClusterStatusUpdateAction is an expected client action to update
+// the status of a cluster.
+type expectedClusterStatusUpdateAction struct {
+	masterNodeGroups  int
+	computeNodeGroups int
+}
+
+func (ea expectedClusterStatusUpdateAction) resource() schema.GroupVersionResource {
+	return clusteroperator.SchemeGroupVersion.WithResource("clusters")
+}
+
+func (ea expectedClusterStatusUpdateAction) verb() string {
+	return "update"
+}
+
+func (ea expectedClusterStatusUpdateAction) validate(t *testing.T, action clientgotesting.Action) bool {
+	if action.GetSubresource() != "status" {
+		return false
+	}
+	updateAction, ok := action.(clientgotesting.UpdateAction)
+	if !ok {
+		t.Errorf("update action is not an UpdateAction: %t", updateAction)
+		return false
+	}
+	updatedObject := updateAction.GetObject()
+	cluster, ok := updatedObject.(*clusteroperator.Cluster)
+	if !ok {
+		t.Errorf("cluster status update action object is not a Cluster: %t", cluster)
+		return true
+	}
+	if e, a := ea.masterNodeGroups, cluster.Status.MasterNodeGroups; e != a {
+		t.Errorf("unexpected masterNodeGroups in cluster update status: expected %v, got %v", e, a)
+	}
+	if e, a := ea.computeNodeGroups, cluster.Status.ComputeNodeGroups; e != a {
+		t.Errorf("unexpected computeNodeGroups in cluster update status: expected %v, got %v", e, a)
+	}
+	return true
+}
+
+// validateControllerExpectations validates that the specified cluster is
+// expecting the specified number of creations and deletions of node groups.
+func validateControllerExpectations(t *testing.T, ctrlr *ClusterController, cluster *clusteroperator.Cluster, expectedAdds, expectedDeletes int) {
+	expectations, ok, err := ctrlr.expectations.GetExpectations(getKey(cluster, t))
+	switch {
+	case err != nil:
+		t.Errorf("error getting expecations: %v", cluster.GetName(), err)
+	case !ok:
+		if expectedAdds != 0 || expectedDeletes != 0 {
+			t.Errorf("no expectations found: expectedAdds %v, expectedDeletes %v", expectedAdds, expectedDeletes)
+		}
+	default:
+		actualsAdds, actualDeletes := expectations.GetExpectations()
+		if e, a := int64(expectedAdds), actualsAdds; e != a {
+			t.Errorf("unexpected number of adds in expectations: expected %v, got %v", e, a)
+		}
+		if e, a := int64(expectedDeletes), actualDeletes; e != a {
+			t.Errorf("unexpected number of deletes in expectations: expected %v, got %v", e, a)
+		}
+	}
+}
+
+// TestSyncClusterSteadyState tests syncing a cluster when it is in a steady
+// state.
+func TestSyncClusterSteadyState(t *testing.T) {
+	cases := []struct {
+		name     string
+		computes []string
+	}{
+		{
+			name: "only master",
+		},
+		{
+			name:     "single compute",
+			computes: []string{"compute1"},
```

Comment:
My understanding is that there will be a single master node group, zero or one infra node groups, and zero or more compute node groups. Is that not correct? I was anticipating that a InfraNodeGroup would be added to the cluster spec.

Reply:
Ahh in current state I thought you were aiming for ComputeNodeGroup housing infra and compute nodes. I'm not sure if this is preferred to treating them more generically. Maybe @csrwng has some ideas.

Reply:
My thinking was that there may be additional work to standup the infra nodes, like there will be to standup the master nodes. So I thought it would be easier to split the infra node group out and give it its own `NodeType`. For what it's worth, I don't have a strong opinion either way: I don't know enough yet about what all needs to be done to standup the infra nodes.

Once @csrwng finishes his work to clean up and improve the cluster-operator API, then this issue will be all sorted out.

Reply:
My thought is yes, we should ultimately allow the user to operate with only a single node group--that being the "master" node group. If there is no "infra" node group, then the "master" node group is used for infra. If there is no "compute" node group, then the "infra" node group (or "master" node group if there is no "infra" node group) is used for compute.

But maybe there is no good use case for that. I don't know.